File: src/main/java/com/qlm/zombie/ai/task/TaskCatalogue.java
/*
 * Copyright (c) 2026 QLM Zombie Mod
 *
 * This file is part of QLM Zombie Mod.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * ----------------------------------------------------------------------------
 * Open Source Attribution:
 *
 * This class is an ORIGINAL implementation inspired by the design patterns of:
 *   - PlayerEngine (https://github.com/Goodbird-git/PlayerEngine)
 *     Copyright (c) Goodbird-git
 *     Licensed under MIT License
 *   - AltoClef TaskCatalogue command-to-task mapping pattern
 *
 * This is NOT a direct copy. The implementation is original Forge 1.20.1 code
 * adapted to the qlmzombie mod's specific requirements.
 * ----------------------------------------------------------------------------
 */
package com.qlm.zombie.ai.task;

import com.qlm.zombie.ai.Player2APIService.AIResponse;
import com.qlm.zombie.entity.FakePlayerEntity;
import net.minecraft.world.entity.player.Player;

/**
 * 任务目录 — 参考 PlayerEngine TaskCatalogue
 * 将 AIResponse (解析后的指令) 映射到具体的 Task 实例
 */
public class TaskCatalogue {

    /**
     * 根据解析后的指令创建任务
     */
    public static Task createTask(FakePlayerEntity ai, Player owner, AIResponse response) {
        switch (response.action()) {
            case "follow" -> { return new FollowTask(ai, owner); }
            case "wait" -> { return new WaitTask(ai, owner); }
            case "come" -> { return new ComeTask(ai, owner); }
            case "stop" -> { return new StopTask(ai, owner); }
            case "mine" -> { return new MineTask(ai, owner); }
            case "chop" -> { return new ChopTask(ai, owner); }
            case "gather" -> { return new GatherTask(ai, owner, response.targetItem()); }
            case "collect_planks" -> { return new CollectPlanksTask(ai, owner); }
            case "attack" -> { return new AttackTask(ai, owner); }
            case "guard" -> { return new GuardTask(ai, owner); }
            case "build", "house" -> { return new BuildTask(ai, owner); }
            case "craft" -> { return new CraftTask(ai, owner, response.targetItem(), response.targetCount()); }
            case "give" -> { return new GiveTask(ai, owner, response.targetItem(), response.targetCount()); }
            case "drop" -> { return new DropTask(ai, owner, response.targetItem(), response.targetCount()); }
            case "explore" -> { return new ExploreTask(ai, owner); }
            case "heal" -> { return new HealTask(ai, owner); }
            case "farm" -> { return new FarmTask(ai, owner); }
            default -> { return null; }
        }
    }
}

File: src/main/kotlin/com/qlm/zombie/ai/task/TaskCatalogue.kt
package com.qlm.zombie.ai.task

import com.qlm.zombie.QLMZombieMod
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.phys.Vec3

object TaskCatalogue {

    private val factories = mutableMapOf<String, (LivingEntity, List<String>) -> Task>()

    init {
        register("mine", "挖矿", "采矿") { entity, args ->
            val radius = args.getOrNull(0)?.toIntOrNull() ?: 5
            val filter = args.getOrNull(1) ?: ""
            MineTask(entity, radius, filter)
        }

        register("chop", "砍树", "伐木") { entity, args ->
            val radius = args.getOrNull(0)?.toIntOrNull() ?: 5
            MineTask(entity, radius, "tree")
        }

        register("collect", "收集", "采集") { entity, args ->
            val radius = args.getOrNull(0)?.toIntOrNull() ?: 5
            val filter = args.getOrNull(1) ?: ""
            MineTask(entity, radius, filter)
        }

        register("attack", "攻击", "战斗") { entity, args ->
            val targetName = args.getOrNull(0) ?: ""
            MineTask(entity, 5, "enemy")
        }

        register("build", "建造", "建房") { entity, args ->
            val size = args.getOrNull(0)?.toIntOrNull() ?: 5
            BuildTask(entity, size)
        }

        register("craft", "制作", "合成") { entity, args ->
            val recipe = args.getOrNull(0) ?: ""
            MineTask(entity, 3, "crafting")
        }

        register("follow", "跟随", "跟着") { entity, args ->
            val playerName = args.getOrNull(0) ?: ""
            val distance = args.getOrNull(1)?.toFloatOrNull() ?: 3f
            FollowTask(entity, playerName, distance)
        }

        register("stop", "停下", "停止") { entity, args ->
            MineTask(entity, 0, "__stop__")
        }
    }

    private fun register(vararg commands: String, factory: (LivingEntity, List<String>) -> Task) {
        for (cmd in commands) {
            factories[cmd.lowercase()] = factory
        }
    }

    fun getTask(command: String, entity: LivingEntity, args: List<String> = emptyList()): Task? {
        val key = command.lowercase().trim()
        val factory = factories[key]
        if (factory == null) {
            QLMZombieMod.LOGGER.warn("[TaskCatalogue] 未知命令: $command")
            return null
        }
        return try {
            factory(entity, args)
        } catch (e: Exception) {
            QLMZombieMod.LOGGER.error("[TaskCatalogue] 创建任务失败: ${e.message}")
            null
        }
    }

    fun getAvailableCommands(): List<String> {
        return factories.keys.toList().sorted()
    }

    fun getCommandHelp(): Map<String, String> {
        return mapOf(
            "挖矿 <半径> [方块名]" to "在指定半径内挖掘方块",
            "砍树 <半径>" to "在指定半径内砍伐树木",
            "收集 <半径> [物品名]" to "在指定半径内收集掉落物",
            "攻击 [目标名]" to "攻击附近的敌对生物",
            "建造 <大小>" to "建造指定大小的小屋",
            "制作 [配方]" to "在工作台制作物品",
            "跟随 [玩家名] [距离]" to "跟随指定玩家",
            "停下" to "停止当前任务"
        )
    }
}